[fix] Resolve group node execution error when connecting to external nodes

christian-byrne opened the pull request:
Fixes a bug where group node children would fail to execute when connected to nodes outside the group, throwing 'Failed to get input node' errors.

The issue was in ExecutableGroupNodeChildDTO.resolveInput which assumed all node IDs follow a colon-separated format, but external nodes have simple IDs. Now properly handles both external connections (trying full ID first) and internal group node connections (falling back to shortened ID).

Includes comprehensive unit tests to prevent regression.

┆Issue is synchronized with this [Notion page](https://www.notion.so/PR-5054-fix-Resolve-group-node-execution-error-when-connecting-to-external-nodes-2526d73d3650816db779fbc080c10579) by [Unito](https://www.unito.io)

christian-byrne committed:
[fix] resolve group node execution error when connecting to external nodes

Fixed ExecutableGroupNodeChildDTO.resolveInput to properly handle connections from group node children to external nodes. The method now tries to find nodes by their full ID first (for external nodes) before falling back to the shortened ID (for internal group nodes).

Added comprehensive unit tests to prevent regression.

AustinMroz commented:
It's an improvement, but things still break if a group node is placed inside a subgraph. Doing some quick testing to see if there's a saner fix than just iteratively trying combinations of id chunks

I'm also seeing errors on execution caused by failing to resolve progress bar display. This seems to be entirely cosmetic and can be pushed to a future fix.

AustinMroz commented:
The issue is more complex than I initially presumed.  While you can calculate the full id of the resolved input from the id of the current ExecutableNodeDTO, it just creates cascading errors when trying to resolve links across the subgraph boundry to the (not actually real) group node.

I think it's fine to merge as is with optional minor nit:
- If a group node is inside a subgraph (I think `this.id.split(':').length > 2` is sufficient?), instead print an error about groupnodes inside subgraphs being unsupported.

christian-byrne commented:
@AustinMroz Thanks for the feedback\! You're right that this doesn't handle group nodes inside subgraphs.

I was thinking we could leverage the existing `parseExecutionId` utility from `graphTraversalUtil.ts` for a more robust solution:

```typescript
// Use parseExecutionId to handle all nesting scenarios
const inputNodeIdParts = parseExecutionId(inputNodeId)
let inputNodeDto = this.nodesByExecutionId?.get(inputNodeId)

if (\!inputNodeDto && inputNodeIdParts) {
  // Try last segment (nodes in same group)
  const lastSegment = inputNodeIdParts[inputNodeIdParts.length - 1]
  inputNodeDto = this.nodesByExecutionId?.get(lastSegment)
  
  // If in subgraph, try constructing proper execution ID
  if (\!inputNodeDto && this.subgraphNodePath.length > 0) {
    const sameSubgraphId = [...this.subgraphNodePath, lastSegment].join(':')
    inputNodeDto = this.nodesByExecutionId?.get(sameSubgraphId)
  }
}
```

This would handle more complex nesting cases. What do you think about this approach vs the iterative solution you mentioned?

christian-byrne committed:
[feat] Add error check for unsupported group nodes inside subgraphs

Added validation to detect when group node children are executing within subgraph contexts (execution ID has >2 segments) and provide clear error message directing users to convert to subgraphs instead.

Includes comprehensive test coverage for the new validation.

christian-byrne commented:
✅ Implemented the suggested check for group nodes inside subgraphs.

Added validation in `resolveInput()` that detects when `this.id.split(':').length > 2` and throws a clear error message directing users to convert to subgraphs instead.

The fix now:
- ✅ Resolves the original group node execution error for main graph
- ✅ Provides clear error for unsupported subgraph nesting 
- ✅ Includes comprehensive unit tests for both scenarios
- ✅ Aligns with the strategic direction to encourage subgraph migration

Ready for review/merge\!
